renderer/reducers/lnd.js

```diff
@@ -429,16 +430,40 @@ export const recoverOldWallet = () => async (dispatch, getState) => {
     // Start lnd with the provided wallet config.
     await dispatch(startLnd(wallet))
 
+    // Set up a listener that resolves once the lightning interface has become active.
+    // This is the point where we really know that the wallet has been created and we can connect to it.
+    const waitForLightning = new Promise(resolve => {
+      handleLightningActive = proxyValue(() => resolve())
+      grpc.prependListener('GRPC_LIGHTNING_SERVICE_ACTIVE', handleLightningActive)
+    })
+
     // Call initWallet method.
-    const grpc = await grpcService
     await grpc.services.WalletUnlocker.initWallet({
       wallet_password: state.onboarding.password,
+      aezeed_passphrase: state.onboarding.passphrase,
       cipher_seed_mnemonic: state.onboarding.seed,
       recovery_window: 2500,
     })
+
+    // Wait for the lightning gRPC interface to become active.
+    await waitForLightning
+
+    // Notify of wallet recovery success.
     dispatch(walletRecovered())
   } catch (error) {
-    dispatch(recoverOldWalletFailure(error))
+    // Attempt to clean up from the failed import attempt.
+    try {
+      const { lndConfig } = getState().lnd
+      await dispatch(stopLnd())
+      await dispatch(removeWallet(lndConfig))
+    } finally {
+      // Remove Lightning grpc activation listener.
+      if (handleLightningActive) {
+        grpc.off('GRPC_LIGHTNING_SERVICE_ACTIVE', handleLightningActive)
```

Comment:
To me it looks like `GRPC_LIGHTNING_SERVICE_ACTIVE` listener will cleanup only if there is an error in the main code, otherwise this `finally` is unreachable

Reply:
This has been refactored, I now use `prependOnceListener` to ensure the callback will only be called once, so if the code block above is successful, there will be nothing to remove so it's fine that this finally block never runs.